### PR TITLE
docs: Documented front-end observability runtime configuration

### PR DESCRIPTION
**Pull request by mcollovati**
Closes vaadin/observability-kit#195

### CODE REVIEW

File: articles/tools/observability/configuration.asciidoc [hunk 1]
```diff
@@ -108,3 +108,125 @@ To enable any of these, add the following line to the [filename]`agent.propertie
...
+
+With the static configuration the same settings are applied to all UI instances and to change the configuration a server restart is required.
+
+Opposite, with dynamic configuration the instrumentation can be enabled or disabled at runtime, without restarting the server, and can be tuned for a specific browser tab. Changes are applied when the browser page is reloaded.
+
+=== Frontend Observability Static Configuration
+
+Static frontend observability configuration can be provided by adding entries to the [filename]`agent.properties` file or with environment variables or system properties, as mentioned earlier.
+
+The following properties can be used to tune frontend instrumentation
```
Comment (jouni) [inline articles/tools/observability/configuration.asciidoc:128]:
```suggestion
The following properties can be used to tune frontend instrumentation:
```

File: articles/tools/observability/configuration.asciidoc [hunk 2]
```diff
@@ -108,3 +108,125 @@ To enable any of these, add the following line to the [filename]`agent.propertie
...
+
+Frontend observability is enabled by default, with all the client side instrumentation active.
+The configuration can be tuned in two ways: 
+
+* statically, by editing the [filename]`agent.properties` or by providing environment variables or system properties
+* dynamically, implementing [interfacename]`ObservabilityClientConfigurer` that applies changes at runtime.
+
+With the static configuration the same settings are applied to all UI instances and to change the configuration a server restart is required.
+
+Opposite, with dynamic configuration the instrumentation can be enabled or disabled at runtime, without restarting the server, and can be tuned for a specific browser tab. Changes are applied when the browser page is reloaded.
```
Comment (jouni) [inline articles/tools/observability/configuration.asciidoc:122]:
```suggestion
With dynamic configuration, you can enable or disable the instrumentation at runtime, without restarting the server, and tune it for a specific browser tab. Changes are applied when the browser page is reloaded.
```

File: articles/tools/observability/configuration.asciidoc [hunk 3]
```diff
@@ -108,3 +108,125 @@ To enable any of these, add the following line to the [filename]`agent.propertie
...
+
+== Frontend Observability Configuration
+
+Frontend observability is enabled by default, with all the client side instrumentation active.
+The configuration can be tuned in two ways: 
+
+* statically, by editing the [filename]`agent.properties` or by providing environment variables or system properties
+* dynamically, implementing [interfacename]`ObservabilityClientConfigurer` that applies changes at runtime.
+
+With the static configuration the same settings are applied to all UI instances and to change the configuration a server restart is required.
```
Comment (jouni) [inline articles/tools/observability/configuration.asciidoc:120]:
```suggestion
With static configuration the same settings are applied to all UI instances. Changes to the configuration require a server restart.
```

File: articles/tools/observability/configuration.asciidoc [hunk 4]
```diff
@@ -108,3 +108,125 @@ To enable any of these, add the following line to the [filename]`agent.propertie
...
 otel.instrumentation.${instrumentationName}.enabled=true
 ----
+
+== Frontend Observability Configuration
+
+Frontend observability is enabled by default, with all the client side instrumentation active.
+The configuration can be tuned in two ways: 
+
+* statically, by editing the [filename]`agent.properties` or by providing environment variables or system properties
+* dynamically, implementing [interfacename]`ObservabilityClientConfigurer` that applies changes at runtime.
```
Comment (jouni) [inline articles/tools/observability/configuration.asciidoc:118]:
```suggestion
* Dynamically, implementing the [interfacename]`ObservabilityClientConfigurer` interface that applies changes at runtime.
```

File: articles/tools/observability/configuration.asciidoc [hunk 5]
```diff
@@ -108,3 +108,125 @@ To enable any of these, add the following line to the [filename]`agent.propertie
 ----
 otel.instrumentation.${instrumentationName}.enabled=true
 ----
+
+== Frontend Observability Configuration
+
+Frontend observability is enabled by default, with all the client side instrumentation active.
```
Comment (jouni) [inline articles/tools/observability/configuration.asciidoc:114]:
```suggestion
Frontend observability is enabled by default, with all the client-side instrumentation active.
```

File: articles/tools/observability/configuration.asciidoc [hunk 6]
```diff
@@ -108,3 +108,125 @@ To enable any of these, add the following line to the [filename]`agent.propertie
...
+=== Frontend Observability Runtime Configuration
+
+By default, frontend observability enables all out-of-the-box client-side instrumentation.
+Observability Kit provides a way to programmatically enable, disable and tune the instrumentation at runtime.
+
+This is done by implementing the [interfacename]`ObservabilityClientConfigurer` interface and applying the desired settings to the [interfacename]`ObservabilityClientConfiguration` instance provided to the [methodname]`configure` method.
+The configuration is applied during UI initialization, meaning that the configurator logic can access the current [classname]`VaadinRequest` and the user details, if a security infrastructure has been configured for the project and the user is logged into the application.
+This allows to fine tune the observability to single user granularity.
+
+For example, the following implementation shows how to set up observability based on a mutable configuration tight to the current logged user.
```
Comment (MatthewVaadin) [inline articles/tools/observability/configuration.asciidoc:174]:
```suggestion
For example, the following implementation shows how to set up observability based on a mutable configuration tied to the currently logged in user.
```

File: articles/tools/observability/configuration.asciidoc [hunk 7]
```diff
@@ -108,3 +108,125 @@ To enable any of these, add the following line to the [filename]`agent.propertie
...
+            return config;
+        }
+        // user not logged-in, return null or a default configuration
+        return null;
+    }
+
+}
+----
+
+With the above implementation, the configuration can for example be fetched from a database table, allowing changes to be applied at runtime after a browser page reload.
```
Comment (MatthewVaadin) [inline articles/tools/observability/configuration.asciidoc:222]:
```suggestion
With the above implementation, the configuration can (for example) be fetched from a database table, allowing changes to be applied at runtime after a browser page reload.
```

File: articles/tools/observability/configuration.asciidoc [hunk 8]
```diff
@@ -108,3 +108,125 @@ To enable any of these, add the following line to the [filename]`agent.propertie
...
+
+The configurator is loaded through the Java ServiceLoader API, adding to [filename]`src/main/resources/META-INF/services/` the file [filename]`com.vaadin.observability.ObservabilityClientConfigurer` with the content being the fully qualified name of the configurator class. For this example, the content would be `org.example.UserBasedFrontendObservability`.
```
Comment (MatthewVaadin) [inline articles/tools/observability/configuration.asciidoc:224]:
```suggestion
The configurator is loaded through the Java ServiceLoader API, by adding the file [filename]`com.vaadin.observability.ObservabilityClientConfigurer` to the [filename]`src/main/resources/META-INF/services/` directory, with the content being the fully qualified name of the configurator class.

For this example, the file looks like this:
`com.vaadin.observability.ObservabilityClientConfigurer`
[source]
----
`org.example.UserBasedFrontendObservability`
----
```

File: articles/tools/observability/configuration.asciidoc [hunk 9]
```diff
@@ -108,3 +108,125 @@ To enable any of these, add the following line to the [filename]`agent.propertie
...
+
+For details on observability client settings, consult the Javadocs of [interfacename]`ObservabilityClientConfiguration` interface.
```
Comment (MatthewVaadin) [inline articles/tools/observability/configuration.asciidoc:226]:
```suggestion
For details on observability client settings, consult the Javadocs of the [interfacename]`ObservabilityClientConfiguration` interface.
```

File: articles/tools/observability/configuration.asciidoc [hunk 10]
```diff
@@ -108,3 +108,125 @@ To enable any of these, add the following line to the [filename]`agent.propertie
...
+
+.Load the configurator in a Spring project
+[NOTE]
+====
+For Spring based project, the configurator can be loaded by exposing the implementation as a Spring managed [annotationname]`@Bean`, without the need to register the class for the ServiceLoader API.
```
Comment (MatthewVaadin) [inline articles/tools/observability/configuration.asciidoc:231]:
```suggestion
For Spring based projects, the configurator can be loaded by exposing the implementation as a Spring managed [annotationname]`@Bean`, without the need to register the class for the ServiceLoader API.
```

**Review by MatthewVaadin — CHANGES_REQUESTED** [review]

**Review by MatthewVaadin — APPROVED** [review]

File: articles/tools/observability/configuration.asciidoc [hunk 11]
```diff
@@ -108,3 +108,125 @@ To enable any of these, add the following line to the [filename]`agent.propertie
...
+
+Frontend observability is enabled by default, with all the client-side instrumentation active.
+The configuration can be tuned in two ways: 
+
+* Statically, by editing the [filename]`agent.properties` file or by providing environment variables or system properties.
+* Dynamically, implementing the [interfacename]`ObservabilityClientConfigurer` interface that applies changes at runtime.
+
+With static configuration the same settings are applied to all UI instances. Changes to the configuration require a server restart.
+
+With dynamic configuration, you can enable or disable the instrumentation at runtime, without restarting the server, and tune it for a specific browser tab. Changes are applied when the browser page is reloaded.
```
[the remark this answers is not in the document]
Comment (jouni) [reply articles/tools/observability/configuration.asciidoc]: I wonder why you wanted to change from active to passive voice in the first sentence? Is it weird to say “you” when it’s the software enabling/disabling the instrumentation based on the configuration?

Splitting it into two standalone sentences is good, though 👍

File: articles/tools/observability/configuration.asciidoc (hunk 7 above)
Comment (jouni) [reply articles/tools/observability/configuration.asciidoc]: I would rather use commas than parenthesis here.

**Pull request merged**